packages/eas-cli/src/utils/code-signing.ts

```diff
@@ -0,0 +1,169 @@
+  validateSelfSignedCertificate(certificate, {
+    publicKey: certificate.publicKey as PKI.rsa.PublicKey,
+    privateKey,
+  });
+
+  return {
+    privateKey,
+    certificate,
+  };
+}
+
+export type MultipartPart = { headers: Map<string, string>; body: string };
+
+export async function parseMultipartMixedResponseAsync(res: Response): Promise<MultipartPart[]> {
+  const contentType = res.headers['content-type'];
+  if (!contentType) {
+    throw new Error('Missing content-type in multipart response');
+  }
+
+  const boundaryRegex = /^multipart\/.+?; boundary=(?:"(.+)"|([^\s]+))$/i;
+  const matches = boundaryRegex.exec(contentType);
+  if (!matches) {
+    throw new Error('content-type header in response not multipart');
+  }
+  const boundary = matches[1] ?? matches[2];
+
+  const bodyBuffer = res.rawBody;
+  const bufferStream = new Stream.PassThrough();
+  bufferStream.end(bodyBuffer);
+
+  return await new Promise((resolve, reject) => {
+    const parts: MultipartPart[] = [];
+    bufferStream.pipe(
+      new Dicer({ boundary })
+        .on('part', p => {
+          const part: MultipartPart = {
+            body: '',
+            headers: new Map(),
+          };
+
+          p.on('header', headers => {
+            for (const h in headers) {
+              part.headers.set(h, (headers as { [key: string]: string[] })[h][0]);
+            }
+          });
+          p.on('data', data => {
+            part.body = data.toString();
+          });
+          p.on('end', () => {
+            parts.push(part);
+          });
+        })
+        .on('finish', () => {
+          resolve(parts);
+        })
+        .on('error', err => reject(err))
+    );
+  });
+}
+
+function isManifestMultipartPart(multipartPart: MultipartPart): boolean {
+  const partName = parseItem(nullthrows(multipartPart.headers.get('content-disposition')))[1].get(
+    'name'
+  );
+  return partName === 'manifest';
+}
+
+export async function getManifestBodyAsync(res: Response): Promise<string | null> {
+  const multipartParts = await parseMultipartMixedResponseAsync(res);
+  const manifestPart = multipartParts.find(isManifestMultipartPart);
+  return manifestPart?.body ?? null;
+}
+
+export function signManifestBody(
+  body: string,
+  certificate: PKI.Certificate,
+  privateKey: PKI.rsa.PrivateKey
+): string {
+  return signStringRSASHA256AndVerify(privateKey, certificate, body);
+}
+
+function assertAssetParity(
+  manifestResponseBodyAssetJSON: { [key: string]: any },
+  partialManifestAsset: PartialManifestAsset
+): void {
+  if (
+    manifestResponseBodyAssetJSON.hash !== partialManifestAsset.fileSHA256 ||
+    manifestResponseBodyAssetJSON.contentType !== partialManifestAsset.contentType ||
+    manifestResponseBodyAssetJSON.key !== partialManifestAsset.bundleKey
+  ) {
+    throw new Error(
+      `Code signing manifest integrity error: Manifest asset tamper detected for asset: ${partialManifestAsset.bundleKey}`
+    );
```

Comment:
If this error occurs and a developer sends the message to us, it will be much more helpful to see which field(s) didn't match and what the expected and actual values were. (Make this message easy for future-us to diagnose.)